programs/prettier.nix

```diff
@@ -0,0 +1,282 @@
+      default = null;
+    };
+    rangeStart = mkOption {
+      description = "Format code starting at a given character offsetypes.\nThe range will extend backwards to the start of the first line containing the selected statementypes.\nThis option cannot be used with --cursor-offsetypes.";
+      type = types.nullOr types.int;
+      example = 0;
+      default = null;
+    };
+    requirePragma = mkOption {
+      description = "Require either '@prettier' or '@format' to be present in the file's first docblock comment\nin order for it to be formatted.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    semi = mkOption {
+      description = "Print semicolons.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    singleAttributePerLine = mkOption {
+      description = "Enforce single attribute per line in HTML, Vue and JSX.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    singleQuote = mkOption {
+      description = "Use single quotes instead of double quotes.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    tabWidth = mkOption {
+      description = "Number of spaces per indentation level.";
+      type = types.nullOr types.int;
+      example = 2;
+      default = null;
+    };
+    trailingComma = mkOption {
+      description = "Print trailing commas wherever possible when multi-line.";
+      type = types.nullOr (types.enum [ "es5" "none" "all" ]);
+      example = "es5";
+      default = null;
+    };
+    useTabs = mkOption {
+      description = "Indent with tabs instead of spaces";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    vueIndentScriptAndStyle = mkOption {
+      description = "Indent script and style tags in Vue files.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    overrides = mkOption {
+      description = "Provide a list of patterns to override prettier configuration.";
+      type = types.nullOr (types.listOf types.attrs);
+      example = { files = [ "*.html" "legacy/**/*.js" ]; options.tabwidth = 4; };
+      default = null;
+    };
+  };
+
+  # We filter out null values in dprint configuration since dprint errors when
+  # they're present
+  mkConfigFile = { filename ? ".prettierrc.json", config ? { } }:
+    configFormat.generate filename (filterAttrsRecursive (n: v: v != null) config);
+
+in
+{
+  options.programs.prettier = {
+    enable = mkEnableOption "prettier";
+    package = mkOption {
+      type = types.package;
+      default = pkgs.nodePackages.prettier;
+      defaultText = lib.literalExpression "pkgs.nodePackages.prettier";
+      description = lib.mdDoc ''
+        prettier derivation to use.
+      '';
+    };
+
+    # Represents the .prettierrc.json config schema
+    config = configSchema;
```

Comment:
"config" is an overloaded term in the nix module system; it's best to use "settings" instead.